core/drivers/crypto/crypto_api/acipher/dsa.c

```diff
@@ -0,0 +1,223 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018-2021 NXP
+ *
+ * Crypto DSA interface implementation to enable HW driver.
+ */
+#include <crypto/crypto.h>
+#include <drvcrypt.h>
+#include <drvcrypt_acipher.h>
+
+/*
+ * Get the recommended L and N bits parameters corresponding
+ * respectively to the size of the Primes P and G (and so
+ * the Public Key and Private Key).
+ *
+ * Refer the NIST.FIPS 186-4 section 4.2
+ *
+ * @size_bits   Maximum key size bits
+ * @l_bits      [out] L size in bits
+ * @n_bits      [out] N size in bits
+ */
+static TEE_Result get_keys_size(size_t size_bits, size_t *l_bits,
+				size_t *n_bits)
+{
+	if (size_bits <= 1024)
+		*n_bits = 160;
+	else if (size_bits <= 3072)
+		*n_bits = 256;
+	else
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
+	*l_bits = size_bits;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result crypto_acipher_alloc_dsa_keypair(struct dsa_keypair *key,
+					    size_t size_bits)
+{
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+	struct drvcrypt_dsa *dsa = NULL;
+	size_t l_bits = 0;
+	size_t n_bits = 0;
+
+	if (!key || !size_bits) {
+		CRYPTO_TRACE("Param error key @0x%" PRIxPTR " size %zu bits",
+			     key, size_bits);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = get_keys_size(size_bits, &l_bits, &n_bits);
+	if (ret == TEE_SUCCESS) {
+		dsa = drvcrypt_get_ops(CRYPTO_DSA);
+		if (dsa)
+			ret = dsa->alloc_keypair(key, l_bits, n_bits);
```

Comment:
shouldnt there be an else and an ret = TEE_ERROR_NOT_IMPLEMENTED?

Reply:
Right, in the current code, we would return the result of ```get_keys_size()``` if dsa == NULL. I will fix it